pkg/storage/stores/shipper/compactor/retention/marker.go

```diff
@@ -34,42 +37,90 @@ type markerStorageWriter struct {
 	tx     *bbolt.Tx
 	bucket *bbolt.Bucket
 
-	count    int64
-	fileName string
+	count            int64
+	currentFileCount int64
+	curFileName      string
+	workDir          string
+
+	buf []byte
 }
 
 func NewMarkerStorageWriter(workingDir string) (MarkerStorageWriter, error) {
-	err := chunk_util.EnsureDirectory(filepath.Join(workingDir, markersFolder))
+	dir := filepath.Join(workingDir, markersFolder)
+	err := chunk_util.EnsureDirectory(dir)
 	if err != nil {
 		return nil, err
 	}
-	fileName := filepath.Join(workingDir, markersFolder, fmt.Sprint(time.Now().UnixNano()))
+
+	msw := &markerStorageWriter{
+		workDir:          dir,
+		currentFileCount: 0,
+		buf:              make([]byte, 8),
+	}
+
+	return msw, msw.createFile()
+}
+
+func (m *markerStorageWriter) createFile() error {
+	fileName := filepath.Join(m.workDir, fmt.Sprint(time.Now().UnixNano()))
 	db, err := shipper_util.SafeOpenBoltdbFile(fileName)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	tx, err := db.Begin(true)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	bucket, err := tx.CreateBucketIfNotExists(chunkBucket)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	level.Info(util_log.Logger).Log("msg", "mark file created", "file", fileName)
+	bucket.FillPercent = 1
+	m.db = db
+	m.tx = tx
+	m.bucket = bucket
+	m.curFileName = fileName
+	m.currentFileCount = 0
+	return nil
+}
+
+func (m *markerStorageWriter) closeFile() error {
+	err := m.tx.Commit()
+	if err != nil {
+		return err
+	}
+	if err := m.db.Close(); err != nil {
+		return err
+	}
+	// The marker file is empty we can remove.
+	if m.currentFileCount == 0 {
+		return os.Remove(m.curFileName)
 	}
-	return &markerStorageWriter{
-		db:       db,
-		tx:       tx,
-		bucket:   bucket,
-		count:    0,
-		fileName: fileName,
-	}, err
+	return nil
 }
 
 func (m *markerStorageWriter) Put(chunkID []byte) error {
-	if err := m.bucket.Put(chunkID, empty); err != nil {
+	if m.currentFileCount > maxMarkPerFile { // roll files when max marks is reached.
+		if err := m.closeFile(); err != nil {
+			return err
+		}
+		if err := m.createFile(); err != nil {
+			return err
+		}
+
+	}
+	// insert in order and fillpercent = 1
+	id, err := m.bucket.NextSequence()
+	if err != nil {
+		return err
+	}
+	binary.BigEndian.PutUint64(m.buf, id) // insert in order using sequence id.
+	if err := m.bucket.Put(m.buf, chunkID); err != nil {
```

Comment:
I think reusing a buf for keys might cause a problem because the comment [here](https://github.com/etcd-io/bbolt/blob/master/bucket.go#L278) says:
`Supplied value must remain valid for the life of the transaction.`
What do you think?

Reply:
the key is cloned, but I'm thinking about the chunkId.

Reply:
Yeah you're right going to keep the fillpercent but revert this change.